### PR TITLE
Primary cache: support reentrancy

### LINKED ISSUE #4947: Deadlock with primary caching is on
I can consistently reproduce a dead-lock on main (as of 4371783) that happens only with primary caching on.

Repro step:
- reset the blueprint:
image

- left space view over middle one:
image

Deadlock occurs then.

Alternative repro:
- start with primary caching off
- enable primary caching

=> deadlock immediately occurs

Stack trace:

image

Repro RRD (800+MB, on private slack):
https://rerunio.slack.com/archives/C041NHU952S/p1706543181xxxx99?thread_ts=1706204242.101919&cid=C041NHU952S

### PR DESCRIPTION
**Pull request by teh-cmc**
This adds support for reentrancy in the latest-at and range caches in order to support a very nasty edge-case where two rayon tasks (i.e. space views) that query the same exact data (e.g. because they are clones of each other) end up running concurrently _on the same thread_.
This can happen because we execute space views through multiple nested layers of parallel iterators, and because rayon's scheduler is a work-stealing one, this effectively means one thread can jump to anywhere in the code at any point, and might do so while holding a lock it shouldn't.
This becomes a problem now that querying data involves mutations and locks, due to the presence of a cache on the path.

There is a lot of complexity we could add on top of what this PR already does in order to make this edge-case more efficient, but there is no reason to go there unless there is any indication that this is good not enough in practice (i.e. you don't even notice it's going on).
If it turns out that we can see glitches in practice, we'll go there.

Taking a step back, it's important to realize that this is just another side-effect of our current "immediate mode querying model", where each space view computes its dataset on its own at the very last second while it is rendering, therefore mixing up computing the data and using the data, running identical queries multiple times, etc.
We already know that we want to --and have to-- move away from this model in order to make our upcoming features possible (on-disk data, component conversions, data overrides, external store hub, etc); so I'd rather not sink any more complexity than the bare minimum required in this thing -- it has to go away anyhow.

- Fixes #4947 

https://github.com/rerun-io/rerun/assets/2910679/b8d8db83-cafb-46ff-ad38-3e41c9d43cd9

https://github.com/rerun-io/rerun/assets/2910679/c3be6250-31ba-4398-9109-356c9b9a8b4e

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4980/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4980/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4980/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4980)
- [Docs preview](https://rerun.io/preview/d6ad8d745eb93f0a8f544b244d82aeffefe26a50/docs) 
- [Examples preview](https://rerun.io/preview/d6ad8d745eb93f0a8f544b244d82aeffefe26a50/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_query_cache/src/latest_at.rs [hunk 1]
```diff
@@ -96,6 +96,11 @@ impl SizeBytes for LatestAtCache {
 }
 
 impl LatestAtCache {
+    #[inline]
+    pub fn is_cached(&self, query: &LatestAtQuery) -> bool {
```
Comment (Wumpf) [inline crates/re_query_cache/src/latest_at.rs:100]: who's actually using this? / what's the envisioned usecase?

File: crates/re_query_cache/src/range.rs [hunk 2]
```diff
@@ -299,20 +299,56 @@ macro_rules! impl_query_archetype_range {
...
+                //
+                // We can add the extra complexity if this proves to be glitchy in real-world
+                // scenarios -- otherwise all of this is giant hack meant to go away anyhow.
+
+                // NOTE: Same logic as what the store does.
+                if query.range.min <= TimeInt::MIN {
+                    let mut reduced_query = query.clone();
+                    // This is the reduced query corresponding to the timeless part of the data.
+                    // It is inclusive and so it will yield `MIN..=MIN` = `[MIN]`.
+                    reduced_query.range.max = TimeInt::MIN; // inclusive
```
Comment (Wumpf) [inline crates/re_query_cache/src/range.rs:321]: I'm a bit confused by this, why was it not needed before and why ca  the query be malformed to begin with

**Review by Wumpf — APPROVED** [review]
great writeups, thank you for diving deep and noting everything down in particular the tradeoffs on ranges.

I'm not sure how much I'd see the operation overall state of things as a hack but you're right that in the future we'll need to do cache filling operations in bulk ahead of time if only to preserve sanity when dealing with promises, so just a matter of wording & taste ;-)

similar, very nit'y note: _technically_ a non-work-stealing job system can also run into this problem if the work scheduled on the worker thread happens to clash with the task it paused. the problem is that it is a scheduler that does work while waiting on task join - it doesn't matter whether it steals that work item from another worker's queue or not which is afaik all work-stealing means here

File: crates/re_query_cache/src/latest_at.rs (hunk 1 above)
Comment (teh-cmc) [reply crates/re_query_cache/src/latest_at.rs:100]: removed

File: crates/re_query_cache/src/range.rs [hunk 3]
```diff
@@ -299,20 +299,56 @@ macro_rules! impl_query_archetype_range {
                     upsert_results::<A, $($pov,)+ $($comp,)*>(arch_views, &mut range_cache.per_data_time)?;
                 }
 
+                Ok(())
+            };
+
+            let iter_callback = |query: &RangeQuery, range_cache: &crate::RangeCache, f: &mut F| -> crate::Result<()> {
+                re_tracing::profile_scope!("range", format!("{query:?}"));
+
+                // We don't bother implementing the slow path here (busy write lock), as that would
```
[the remark this answers is not in the document]
Comment (teh-cmc) [reply crates/re_query_cache/src/range.rs:308]: That is way too spammy in practice (i tried) -- missing the lock happens quite literally all the time if you have a handful of views asking for the same data as long as they run in parallel.

What's really important though is whether the locked was missed _and_ the data wasn't cached by neither the current thread nor any other; but that we cannot really know for the range case unless we add some more complexity.
Though if it does happen in practice, we should be noticing it in the form of awful visual glitches pretty quickly.

File: crates/re_query_cache/src/range.rs (hunk 2 above)
Comment (Wumpf) [reply crates/re_query_cache/src/range.rs:321]: thanks for the link!

**Pull request merged**